service/matching/db.go

```diff
@@ -165,13 +170,23 @@ func (db *taskQueueDB) renewTaskQueueLocked(
 	return nil
 }
 
-// UpdateState updates the queue state with the given value
+// UpdateState updates the queue state with the given values
 func (db *taskQueueDB) UpdateState(
 	ctx context.Context,
 	ackLevel int64,
 ) error {
 	db.Lock()
 	defer db.Unlock()
+
+	// TODO Shivam: Do we want to fetch the latest ackLevel from taskAckManager?
+	//ackLevel := db.backlogMgr.taskAckManager.getAckLevel() // latest value of ackLevel
+
+	maxReadLevel := db.backlogMgr.taskWriter.GetMaxReadLevel()
```

Comment:
what if we read maxReadLevel after [this](https://github.com/temporalio/temporal/blob/56bbc1443fba7583bed2fbc5c37675fc32eeeb14/service/matching/task_writer.go#L247) appendTasks and before the `atomic.StoreInt64(&w.maxReadLevel, maxReadLevel)` statement in task_writer? wouldn't we reset to zero wrongly?

Reply:
yeah, thanks for the catch. Think it would, since the two operations occurring concurrently could lead to a race condition. Think the solution will be to add a lock of some sorts over here.

Reply:
It was found out that the aforementioned solution would result in a deadlock situation. To solve this, `max_read_level` was moved inside of `db` and the problem has been resolved.